<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>CMAPLE: What&#39;s CMAPLE library?</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">CMAPLE<span id="projectnumber">&#160;1</span>
   </div>
   <div id="projectbrief">CMaple phylogenetic software</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">What's CMAPLE library? </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="md__c_maple_lib"></a> CMAPLE is a C++ reimplementation of <a href="https://www.nature.com/articles/s41588-023-01368-0">MAPLE</a> - a novel likelihood-based phylogenetic inference method for pandemic-scale epidemiological genomic data. CMAPLE is highly optimized for performance and scalability with many new features.</p>
<p>CMAPLE library provides a set of APIs, which allow users to integrate CMAPLE into existing phylogenetic inference methods.</p>
<h1><a class="anchor" id="autotoc_md1"></a>
How to use?</h1>
<p>In the following, we give an instruction of how to include CMAPLE library into a project using CMake build system (instructions for other build systems will be updated soon). Here, we'll take <a href="https://github.com/iqtree/iqtree2">IQ-TREE</a> as an example.</p>
<h2><a class="anchor" id="autotoc_md2"></a>
Include CMAPLE as a submodule of your project</h2>
<p>In your project directory, run </p><pre class="fragment">    git submodule add https://github.com/iqtree/cmaple.git
</pre><h2><a class="anchor" id="autotoc_md3"></a>
Update CMakeList to include CMAPLE</h2>
<pre class="fragment">    project(iqtree)

    # Step 1: Add cmaple directory to the build
    add_subdirectory(cmaple)
    # add the main and other directories of your project
    add_subdirectory(main)

    # Step 2: Build two executables for DNA and protein data
    # build an executable for DNA data
    add_executable(iqtree2
        main.cpp main.h
    )
    # build another executable for protein data
    add_executable(iqtree2-aa
        main.cpp main.h
    )

    # Step 3: Add the binary tree to the search path for include files so that we can find cmaple/cmaple_config.h
    include_directories("${PROJECT_BINARY_DIR}/cmaple")

    # Step 4: Add linking libraries
    target_link_libraries(iqtree2 main maple)
    target_link_libraries(iqtree2-aa main maple-aa)
</pre> <h2><a class="anchor" id="autotoc_md4"></a>
An example of APIs usage</h2>
<pre class="fragment">    #include “cmaple.h”

    // Create an alignment from a file
    cmaple::Alignment aln("alignment.maple");

    // Check if the input alignment is suitable for using [C]Maple method
    if (cmaple::checkMapleSuitability(aln))
    {
        // Create a default model according to the data type from the alignment (i.e., GTR for DNA, and LG for protein data)
        cmaple::Model model(cmaple::ModelBase::DEFAULT, aln.getSeqType());

        // Create a tree, attach the alignment and model to the tree
        cmaple::Tree tree(&amp;aln, &amp;model);

        // Infer a phylogenetic tree from the alignment and the model using [C]Maple algorithm
        tree.autoProceedMAPLE();

        // Compute the branch supports for the inferred tree
        tree.computeBranchSupport();

        // Compute the likelihood of the tree
        cout &lt;&lt; "- Tree log likelihood: " &lt;&lt; tree.computeLh() &lt;&lt; endl;

        // Export the tree (with branch supports) in NEWICK format
        cout &lt;&lt; "- Tree: " &lt;&lt; tree.exportNewick("cmaple::Tree::BIN_TREE, true) &lt;&lt; endl;
    }
    else
    {
        // Execute existing methods (e.g., IQ-TREE, RAXML, PHYML) to analyse this alignment
    }
</pre> <h2><a class="anchor" id="autotoc_md5"></a>
Tips for Debugging</h2>
<p>CMAPLE outputs debugging messages to the standard output <code>std::cout</code>. One could control the amount of those messages via setting <code>cmaple::verbose_mode</code> to one of the following values. <br  />
 - <code>VB_QUIET</code>: no messages except errors. <br  />
 - <code>VB_MED</code> (default): common messages (showing the processing progress). <br  />
 - <code>VB_DEBUG</code>: as many messages as possible (useful for debugging).</p>
<p><br  />
 </p>
<h1><a class="anchor" id="autotoc_md6"></a>
More APIs?</h1>
<p>CMAPLE APIs are exposed in the following. <br  />
<a href="group__cmaple.html"><b>CMaple</b></a> <br  />
<a href="classcmaple_1_1_alignment.html"><b>Alignment</b></a> <br  />
<a href="classcmaple_1_1_model.html"><b>Model</b></a> <br  />
<a href="classcmaple_1_1_tree.html"><b>Tree</b></a></p>
<p><br  />
 </p>
<h1><a class="anchor" id="autotoc_md7"></a>
How to cite CMAPLE?</h1>
<p>To be updated... </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8
</small></address>
</body>
</html>
